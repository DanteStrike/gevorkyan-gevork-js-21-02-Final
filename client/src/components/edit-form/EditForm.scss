.edit-form {
  max-width: 500px;
  padding: 10px;
  margin: 0 auto;
  background-color: var(--primary-background);
  border-radius: 15px;

  &__last {
    margin-bottom: 0;
  }

  &__date-picker {
    width: 100%;
  }

  &__avatar {
    display: flex;
    justify-content: center;
  }

  .ant-upload.ant-upload-select-picture-card {
    padding: 2px;
    margin: 0;
    border-width: 2px;
    border-radius: 50%;
  }

  .ant-form-item-label {
    padding: 0;
    font-weight: 700;
  }

  .ant-form-item-label > label {
    color: var(--primary-font-color);
  }
}
