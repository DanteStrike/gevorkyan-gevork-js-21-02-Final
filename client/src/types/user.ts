export type UserTitleType = 'mr' | 'ms' | 'mrs' | 'miss' | 'dr' | '';
export type UserGenderType = 'male' | 'female' | 'other' | '';

export interface IUserPreview {
  id: string;
  title?: UserTitleType;
  firstName: string;
  lastName: string;
  picture?: string;
}

export interface ILocation {
  street: string;
  city: string;
  state: string;
  country: string;
  timezone: string;
}

export interface IUser extends IUserPreview {
  gender: UserGenderType;
  email: string;
  dateOfBirth: string;
  registerDate: string;
  phone: string;
  location: ILocation;
}

export interface IUserRegistration {
  firstName: string;
  lastName: string;
  gender: UserGenderType;
  dateOfBirth: string;
  email: string;
  phone: string;
}

export interface IUserUpdate extends Omit<IUserRegistration, 'email'> {
  id: string;
  picture: string;
}
