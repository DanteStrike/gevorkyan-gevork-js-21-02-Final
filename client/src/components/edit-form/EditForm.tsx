import React, {useEffect, useMemo, useState} from 'react';
import moment, {Moment} from 'moment';
import {Form, Input, Select, DatePicker, Upload, Avatar} from 'antd';
import Button from '../submit-button/SubmitButton';
import './EditForm.scss';
import {ValidateUtils, DataUtils, FileUtils, ObjectUtils, RequestUtils} from '../../utils';
import {IUser, IUserUpdate, UserGenderType} from '../../types';
import UploadButton from '../upload-button/UploadButton';
import useAppTranslation from '../../hooks/use-app-translation';

interface IEditFormProps {
  user: IUser;
  loading?: boolean;
  onSubmit?: (data: IUserUpdate) => void;
}
interface IEditFormState {
  loading: boolean;
  img: string | null;
}
interface IEditForm {
  name: string;
  gender: UserGenderType;
  dateOfBirth: Moment;
  phone: string;
}

const normFile = (e: any) => {
  if (Array.isArray(e)) {
    return e;
  }
  return e && e.fileList;
};
const avatarStyle = {width: '100%', height: '100%'};
const {Item} = Form;

function EditForm({user, onSubmit = () => {}, loading}: IEditFormProps) {
  const {t} = useAppTranslation(`editForm`);
  const [avatar, setAvatar] = useState<IEditFormState>({
    loading: false,
    img: user.picture || null,
  });
  const handleAvatarChange = (info: any) => {
    FileUtils.encodeToBase64(info.file.originFileObj, (res) => {
      if (typeof res === 'string' || res instanceof String) {
        setAvatar(
          ObjectUtils.updateObject(avatar, {
            loading: false,
            img: res as string,
          })
        );
      }
    });
  };

  const [form] = Form.useForm();
  const defaultValues = useMemo(
    () => ({
      name: DataUtils.collectName(user.firstName, user.lastName),
      picture: user.picture || ``,
      gender: user.gender,
      dateOfBirth: moment(user.dateOfBirth),
      phone: user.phone,
    }),
    [user]
  );

  useEffect(() => {
    form.setFieldsValue(defaultValues);
    setAvatar((prev) => ObjectUtils.updateObject(prev, {img: defaultValues.picture}));
  }, [form, defaultValues]);

  const handleFormFinish = (filedValues: IEditForm) => {
    const normName = DataUtils.normalizeName(filedValues.name);
    const userUpdate: IUserUpdate = {
      id: user.id,
      firstName: normName.firstName,
      lastName: normName.lastName,
      picture: avatar.img || user.picture || ``,
      gender: filedValues.gender,
      dateOfBirth: filedValues.dateOfBirth.toISOString(),
      phone: filedValues.phone,
    };

    onSubmit(userUpdate);
  };

  return (
    <Form
      className="edit-form"
      form={form}
      name="edit"
      size="middle"
      layout="vertical"
      validateTrigger="onSubmit"
      onFinish={handleFormFinish}
      initialValues={defaultValues}
    >
      <Item name="picture" label={t(`avatar`)} valuePropName="file" getValueFromEvent={normFile} noStyle>
        <Upload
          name="avatar"
          listType="picture-card"
          className="edit-form__avatar"
          showUploadList={false}
          accept="image/*"
          beforeUpload={ValidateUtils.imageValidator}
          onChange={handleAvatarChange}
          customRequest={RequestUtils.dummyRequest}
        >
          {avatar.img ? <Avatar src={avatar.img} style={avatarStyle} /> : <UploadButton loading={avatar.loading} />}
        </Upload>
      </Item>

      <Item
        name="name"
        label={t(`name.title`)}
        rules={[ValidateUtils.requireValidatorName, ValidateUtils.userNameValidator]}
      >
        <Input placeholder={t(`name.placeholder`)} />
      </Item>

      <Item name="gender" label={t(`gender.title`)} rules={[ValidateUtils.requireValidatorGender]}>
        <Select>
          <Select.Option value="male">{t(`gender.options.male`)}</Select.Option>
          <Select.Option value="female">{t(`gender.options.female`)}</Select.Option>
          <Select.Option value="other">{t(`gender.options.other`)}</Select.Option>
        </Select>
      </Item>

      <Item
        name="dateOfBirth"
        label={t(`dateOfBirth.title`)}
        rules={[ValidateUtils.requireValidatorBirth, ValidateUtils.sevenYearsOldValidator]}
      >
        <DatePicker
          className="edit-form__date-picker"
          placeholder={t(`dateOfBirth.placeholder`)}
          disabledDate={ValidateUtils.disableDateOverCurrent}
        />
      </Item>

      <Item
        name="phone"
        label={t(`phone.title`)}
        rules={[ValidateUtils.requireValidatorPhone, ValidateUtils.phoneFormatValidator()]}
      >
        <Input />
      </Item>

      <Item className="edit-form__last">
        <Button loading={loading}>{t(`save`)}</Button>
      </Item>
    </Form>
  );
}

EditForm.defaultProps = {
  loading: false,
  onSubmit: () => {},
};

export default React.memo(EditForm);
