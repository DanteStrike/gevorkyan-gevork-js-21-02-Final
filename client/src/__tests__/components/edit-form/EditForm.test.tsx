import {mount, shallow} from 'enzyme';
import React from 'react';
import moment from 'moment';
import EditForm from '../../../components/edit-form/EditForm';
import {fullUser} from '../../mocks/user.mock';
import sleep from '../../mocks/utils';

describe(`Component Footer should render correctly`, () => {
  const user = {...fullUser, dateOfBirth: `1996-04-30T19:26:49.610Z`};

  it(`snapshot default test`, () => {
    const wrapper = shallow(<EditForm user={user} />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`snapshot loading test`, () => {
    const wrapper = shallow(<EditForm user={user} loading />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component Footer should work correctly`, () => {
  const user = {...fullUser, phone: `+79999999999`, dateOfBirth: `1996-04-30T19:26:49.610Z`};

  it(`call submit test test`, () => {
    const submitHandle = jest.fn();
    const wrapper = mount(<EditForm user={user} onSubmit={submitHandle} />);

    wrapper
      .children()
      .prop(`form`)
      .setFieldsValue({...user, dateOfBirth: moment(`1996-04-30T19:26:49.610Z`)});
    wrapper.update();
    wrapper.find('form').simulate('submit');

    return sleep().then(() => {
      expect(submitHandle).toHaveBeenCalledTimes(1);
    });
  });
});
