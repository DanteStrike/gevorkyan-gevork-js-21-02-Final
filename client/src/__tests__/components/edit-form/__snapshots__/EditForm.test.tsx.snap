// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Footer should render correctly snapshot default test 1`] = `
<ForwardRef(InternalForm)
  className="edit-form"
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  initialValues={
    Object {
      "dateOfBirth": "1996-04-30T19:26:49.610Z",
      "gender": "female",
      "name": "anyFN antLN",
      "phone": "phone",
      "picture": "url",
    }
  }
  layout="vertical"
  name="edit"
  onFinish={[Function]}
  size="middle"
  validateTrigger="onSubmit"
>
  <FormItem
    getValueFromEvent={[Function]}
    label="key 'editForm.avatar (en-US)' returned an object instead of string."
    name="picture"
    noStyle={true}
    valuePropName="file"
  >
    <Upload
      accept="image/*"
      action=""
      beforeUpload={[Function]}
      className="edit-form__avatar"
      customRequest={[Function]}
      data={Object {}}
      disabled={false}
      listType="picture-card"
      multiple={false}
      name="avatar"
      onChange={[Function]}
      showUploadList={false}
      supportServerRender={true}
      type="select"
    >
      <Avatar
        shape="circle"
        size="default"
        src="url"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </Upload>
  </FormItem>
  <FormItem
    label="Full name:"
    name="name"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Input
      placeholder="Input full name"
      type="text"
    />
  </FormItem>
  <FormItem
    label="Gender:"
    name="gender"
    rules={
      Array [
        [Function],
      ]
    }
  >
    <ForwardRef(InternalSelect)>
      <Option
        value="male"
      >
        Male
      </Option>
      <Option
        value="female"
      >
        Female
      </Option>
      <Option
        value="other"
      >
        Other
      </Option>
    </ForwardRef(InternalSelect)>
  </FormItem>
  <FormItem
    label="Date of birth:"
    name="dateOfBirth"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Picker
      className="edit-form__date-picker"
      disabledDate={[Function]}
      placeholder="Choose date"
    />
  </FormItem>
  <FormItem
    label="Phone:"
    name="phone"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct phone",
          "pattern": /\\^\\\\\\+\\?7\\(\\\\d\\{10\\}\\)\\$/,
        },
      ]
    }
  >
    <Input
      type="text"
    />
  </FormItem>
  <FormItem
    className="edit-form__last"
  >
    <Memo(SubmitButton)
      loading={false}
    >
      Save
    </Memo(SubmitButton)>
  </FormItem>
</ForwardRef(InternalForm)>
`;

exports[`Component Footer should render correctly snapshot loading test 1`] = `
<ForwardRef(InternalForm)
  className="edit-form"
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  initialValues={
    Object {
      "dateOfBirth": "1996-04-30T19:26:49.610Z",
      "gender": "female",
      "name": "anyFN antLN",
      "phone": "phone",
      "picture": "url",
    }
  }
  layout="vertical"
  name="edit"
  onFinish={[Function]}
  size="middle"
  validateTrigger="onSubmit"
>
  <FormItem
    getValueFromEvent={[Function]}
    label="key 'editForm.avatar (en-US)' returned an object instead of string."
    name="picture"
    noStyle={true}
    valuePropName="file"
  >
    <Upload
      accept="image/*"
      action=""
      beforeUpload={[Function]}
      className="edit-form__avatar"
      customRequest={[Function]}
      data={Object {}}
      disabled={false}
      listType="picture-card"
      multiple={false}
      name="avatar"
      onChange={[Function]}
      showUploadList={false}
      supportServerRender={true}
      type="select"
    >
      <Avatar
        shape="circle"
        size="default"
        src="url"
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </Upload>
  </FormItem>
  <FormItem
    label="Full name:"
    name="name"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Input
      placeholder="Input full name"
      type="text"
    />
  </FormItem>
  <FormItem
    label="Gender:"
    name="gender"
    rules={
      Array [
        [Function],
      ]
    }
  >
    <ForwardRef(InternalSelect)>
      <Option
        value="male"
      >
        Male
      </Option>
      <Option
        value="female"
      >
        Female
      </Option>
      <Option
        value="other"
      >
        Other
      </Option>
    </ForwardRef(InternalSelect)>
  </FormItem>
  <FormItem
    label="Date of birth:"
    name="dateOfBirth"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Picker
      className="edit-form__date-picker"
      disabledDate={[Function]}
      placeholder="Choose date"
    />
  </FormItem>
  <FormItem
    label="Phone:"
    name="phone"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct phone",
          "pattern": /\\^\\\\\\+\\?7\\(\\\\d\\{10\\}\\)\\$/,
        },
      ]
    }
  >
    <Input
      type="text"
    />
  </FormItem>
  <FormItem
    className="edit-form__last"
  >
    <Memo(SubmitButton)
      loading={true}
    >
      Save
    </Memo(SubmitButton)>
  </FormItem>
</ForwardRef(InternalForm)>
`;
